### PR TITLE
KAFKA-15511: Catch CorruptIndexException instead of CorruptRecordException 

### PR DESCRIPTION
**Pull request by iit2009060**
**Scenario**
The bug occurs  when RemoteIndexCache does not have an entry but CorruptedIndex File exists in the desired path. 
**Current behaviour** 
It will throw the CorruptIndex Exception and does not recover or replace the CorruptedIndexFile with new File fetch from remote storage.
 **Desired Behaviour**
It should catch the CorruptIndex Exception and should refetch it  from remote storage and overwrite the corrupted file.

**Testing** 
Written Unit Test to test the above scenario

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: nit

spaces after `,` for each argument

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:621]: please add a comment on why this is considered as "corrupted"

**Review by divijvaidya — COMMENTED** [review]
Looks good. Left some minor comments.

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (iit2009060) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:621]: @divijvaidya  done

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (iit2009060) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: done

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: You will notice that checkstyle is failing in the build (you can check the CI by clicking the "details" button below) since we still didn't add space after each comma.

```
* What went wrong:

Execution failed for task ':storage:checkstyleMain'.

> A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction

   > Checkstyle rule violations were found. See the report at: file:///home/jenkins/workspace/Kafka_kafka-pr_PR-14459/storage/build/reports/checkstyle/main.html

     Checkstyle files with violations: 1

     Checkstyle violations by severity: [error:1]
```

I would suggest to run `./gradlew unitTest` locally to ensure that build is successful before updating the PR.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java
Comment (Hangleton) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:313]: Can `CorruptRecordException` be thrown in some scenario?

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (iit2009060) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: Yes @divijvaidya I will do it from next commit , but is there a way to open this file , it seems to be created in build server ?
"file:///home/jenkins/workspace/Kafka_kafka-pr_PR-14459/storage/build/reports/checkstyle/main.html"

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: No, unfortunately not. The best way to get the checkstyle report is to run it locally.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java
Comment (iit2009060) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:313]: @Hangleton  The  RemoteIndexCache  caches only index files which needs to be checked for corruption. There is no such record concept in RemoteIndexCache. 
@divijvaidya  @satishd can add more details here.

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (iit2009060) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: @divijvaidya  fixed the build issue.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java
Comment (Hangleton) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:313]: Thanks, I wanted to confirm that it is indeed the case although we are dealing with indexes here. The rationale being that `CorruptRecordException` may have been introduced purposefully to cover a use case. Was this not covered by a test in the initial PR, and if it was, did the test make an incorrect assumption, or exercised an invalid (that is, impossible) scenario?

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (satishd) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: @iit2009060 You can run the below command to catch different errors related to check-style, license header checks etc and also runs on different scala versions. It is good to run it before raising PR to catch errors. 

```
./gradlewAll clean releaseTarGz checkStyleMain checkStyleTest spotBugsMain spotBugsTest rat
```

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (satishd) [reply core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: @iit2009060 You can run the below command to catch different errors related to check-style, license header checks etc and also runs on different scala versions. It is good to run it before raising PR to catch errors. 

```
./gradlewAll clean releaseTarGz checkStyleMain checkStyleTest spotBugsMain spotBugsTest rat
```

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java
Comment (divijvaidya) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:313]: > Was this not covered by a test in the initial PR

@Hangleton no we don't have a test case to execute this code. In fact, the bug fixed this PR was discovered while working on the task [1] to add more unit test for remote index cache.

[1] https://issues.apache.org/jira/browse/KAFKA-15169

**Comment by divijvaidya** [comment]
Test failures are unrelated to Tiered Storage code.
```
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.NamedTopologyIntegrationTest.shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.streams.integration/NamedTopologyIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest.testBrokerCoordinator](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.connect.integration/ConnectWorkerIntegrationTest/Build___JDK_11_and_Scala_2_13___testBrokerCoordinator/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.connect.integration/ConnectorRestartApiIntegrationTest/Build___JDK_11_and_Scala_2_13___testMultiWorkerRestartOnlyConnector/)
[Build / JDK 11 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testNoConsumeWithDescribeAclViaAssign(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_11_and_Scala_2_13___testNoConsumeWithDescribeAclViaAssign_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestExcludingClusterAuthorizedOperations(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.server/DescribeClusterRequestTest/Build___JDK_11_and_Scala_2_13___testDescribeClusterRequestExcludingClusterAuthorizedOperations_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_11_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testReplicateSourceDefault()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_8_and_Scala_2_12___testReplicateSourceDefault__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testReplicateSourceDefault()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_8_and_Scala_2_12___testReplicateSourceDefault___2/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testOffsetSyncsTopicsOnTarget()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsWithCustomForwardingAdminIntegrationTest/Build___JDK_8_and_Scala_2_12___testOffsetSyncsTopicsOnTarget__/)
[Build / JDK 8 and Scala 2.12 / kafka.admin.TopicCommandIntegrationTest.testDescribeUnderMinIsrPartitions(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.admin/TopicCommandIntegrationTest/Build___JDK_8_and_Scala_2_12___testDescribeUnderMinIsrPartitions_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testNoGroupAcl(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_8_and_Scala_2_12___testNoGroupAcl_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testDescribeTokenForOtherUserFails(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_8_and_Scala_2_12___testDescribeTokenForOtherUserFails_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / kafka.api.GroupEndToEndAuthorizationTest.testProduceConsumeTopicAutoCreateTopicCreateAcl(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/kafka.api/GroupEndToEndAuthorizationTest/Build___JDK_8_and_Scala_2_12___testProduceConsumeTopicAutoCreateTopicCreateAcl_String__quorum_zk/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/6/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
```

**Review by divijvaidya — APPROVED** [review]
Change looks good to me.

**Comment by divijvaidya** [comment]
@Hangleton @satishd since you folks participated in the review for this PR, do you have any additional comments?

**Review by satishd — APPROVED** [review]

File: core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala]: nit: space after comma and remove extra brackets

**Review by kamalcph — APPROVED** [review]
LGTM, thanks for fixing this issue!

Do you know the reason behind `CorruptIndexException` directly extends the `RuntimeException` instead of `KafkaException`?

**Comment by divijvaidya** [comment]
Unrelated test failures
```
[Build / JDK 8 and Scala 2.12 / integration.kafka.server.FetchFromFollowerIntegrationTest.testFollowerCompleteDelayedFetchesOnReplication(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_8_and_Scala_2_12___testFollowerCompleteDelayedFetchesOnReplication_String__quorum_kraft/)
[Build / JDK 21 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testDescribeTokenForOtherUserPasses(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_21_and_Scala_2_13___testDescribeTokenForOtherUserPasses_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/kafka.api/TransactionsTest/Build___JDK_11_and_Scala_2_13___testBumpTransactionalEpoch_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testNoOpRecordWriteAfterTimeout()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testNoOpRecordWriteAfterTimeout__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14459/7/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_17_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
```

**Review by divijvaidya — APPROVED** [review]

**Comment by iit2009060** [comment]
@divijvaidya  It seems i don't have permission to merge it even after the approval.

**Comment by divijvaidya** [comment]
Merged to trunk and 3.6.

Thank you for the contribution @iit2009060. For Apache projects, only the committers can merge code into the repository.

**Comment by showuon** [comment]
Thanks for the fix @iit2009060 ! Nice find!

**Pull request merged**
